src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -219,19 +219,30 @@
   "Initializes a 'worker' which will listen for lifecycle-related tasks and perform
   them on a background thread, to ensure that we aren't executing multiple lifecycle
   tasks simultaneously."
-  [lifecycle-chan :- (schema/protocol async-prot/Channel)
+  [lifecycle-channel :- (schema/protocol async-prot/Channel)
+   shutdown-channel :- (schema/protocol async-prot/Channel)
    shutdown-reason-promise :- IDeref]
+  (log/debug "Initializing lifecycle worker loop.")
   (async/go (try
               (loop []
-                (let [task (async/<! lifecycle-chan)]
+                (let [[task chan] (async/alts!
+                                   [shutdown-channel lifecycle-channel]
+                                   :priority true)]
                   (schema/validate LifeCycleTask task)
                   (let [{:keys [type task-function]} task]
                     (condp #(contains? %1 %2) type
                       #{:shutdown}
                       (do
                         (log/debug "Received shutdown command, shutting down services")
+                        (async/close! shutdown-channel)
+                        (async/close! lifecycle-channel)
                         (task-function)
                         (log/debug "Service shutdown complete, exiting lifecycle worker loop")
+
+                        ;; drain the channels to ensure that there are
+                        ;; no blocking puts
+                        (while (async/poll! shutdown-channel))
+                        (while (async/poll! lifecycle-channel))
```

Comment:
Cool, I might add "since these channels were just closed above" to the end of that comment, or something like that.